aspnet/security/2fa.rst

```diff
@@ -1,189 +1,171 @@
+After you create the project, follow the instructions in :doc:`accconfirm` to set up and require SSL.
 
-At this point, the project is created. It may take a few moments to load, and you may notice Visual Studio's status bar indicates that Visual Studio id downloading some resources as part of this process.  Visual Studio ensures some required files are pulled into the project when a solution is opened (or a new project is created), and other files may be pulled in at compile time.
+Setup up SMS for two-factor authentication with Twilio
+-------------------------------------------------------
 
+ - Create a `Twilio <http://www.twilio.com/>`_ account.
+ - On the **Dashboard** tab of your Twilio account, note the **Account SID** and **Authentication token**. Note: Tap **Show API Credentials** to see the Authentication token.
+ - On the **Numbers** tab, note the Twilio phone number.
+ - Install the Twilio NuGet package. From the Package Manager Console (PMC),  enter the following the following command:
 
-Running the Application
------------------------
+ ``Install-Package Twilio``
+  - Add code in the *Services/MessageServices.cs* file to enable SMS.
 
-Run the application and after a quick build step, you should see it open in your web browser.
+.. literalinclude:: 2fa/sample/WebSMS/src/WebSMS/Services/MessageServices.cs
+   :language: c#
+   :lines: 12-39
+   :dedent: 4
 
-.. image:: 2fa/_static/first-run.png
+.. comment note:: Twilio cannot target dnxcore50: You will get compilation errors if you build your project when dnxcore50 is included because Twilio does not have a package for dnxcore50. You can remove dnxcore50 from the *project.json* file or you can call the Twilio REST API to send SMS messages.
 
+Configure the  SMS provider key/value
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Setup up SMS for Two-factor authentication
-------------------------------------------
-This tutorial uses Twilio, but you can use any SMS provider. 
+We'll use the :ref:`Options pattern <options-config-objects>` to access the user account and key settings. For more information, see :doc:`../fundamentals/configuration`.
 
- - Create a `Twilio <http://www.twilio.com/>`_ account.
- - From the Dashboard tab of your Twilio account, copy the Account SID and Auth token.
- - From the Numbers tab, copy your Twilio phone number.
- - Make the SID, account token and phone number available to the app:
- - Add the Twilio NuGet package to the app.
+ - Create a class to fetch the secure SMS key. For this sample, the ``AuthMessageSMSSenderOptions`` class is created in the *Services/AuthMessageSMSSenderOptions.cs* file.
 
- - Add code in MessageServices to send SMS
+.. literalinclude:: 2fa/sample/WebSMS/src/WebSMS/Services/AuthMessageSMSSenderOptions.cs
+   :language: c#
+   :lines: 3-8
+   :dedent: 4
 
-.. code-block:: c#
+Set ``SID``, ``AuthToken``, and ``SendNumber`` with the `secret-manager tool <http://docs.asp.net/en/latest/security/app-secrets.html>`_. For example:
```

Comment:
Recommend being more explicit about setting these values to the values obtained from Twilio.